API: added array

TomAugspurger opened the pull request:
Adds

- a new top-level `pd.array` method for creating arrays
- all our extension dtypes to the top-level API
- adds pd.arrays for all our EAs

TODO

- [x] Add the actual array classes somewhere to the public API (pandas.arrays?)
- [x] API docs for the rest of the arrays and dtypes.

Closes #22860 

supersedes #23532.

TomAugspurger committed:
added array

pep8speaks commented:
Hello @TomAugspurger! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/api.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/core/api.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/__init__.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/core/arrays/__init__.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/array_.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/core/arrays/array_.py) !

 - There are no PEP8 issues in the file [`pandas/core/arrays/interval.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/core/arrays/interval.py) !

 - There are no PEP8 issues in the file [`pandas/core/dtypes/dtypes.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/core/dtypes/dtypes.py) !

 - There are no PEP8 issues in the file [`pandas/tests/api/test_api.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/tests/api/test_api.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arrays/test_array.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/tests/arrays/test_array.py) !

 - There are no PEP8 issues in the file [`pandas/tests/arrays/test_period.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/tests/arrays/test_period.py) !

 - There are no PEP8 issues in the file [`pandas/tests/extension/decimal/array.py`](https://github.com/TomAugspurger/pandas/blob/bfefc96913a0bafeb19172b95c30e5dc09d4ad48/pandas/tests/extension/decimal/array.py) !

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
update registry test

TomAugspurger committed:
update doc examples

TomAugspurger commented:
I'd like to restructure the documentation a bit. I want to collect all of our custom types under a new heading like "Pandas Arrays and Types". Under that we would have

- Categorical
  - dtype
  - Categorical
  - CategoricalIndex?
- Period
  - dtype
  - scalar
  - array
  - PeriodIndex(?) or leave that under indexes probably
- Integer
  - ...
- Interval
  - ...
- Sparse
  - ...

and then eventually DatetimeArray & TimedeltaArray. Think of this as "grouping by topic".

Alternatively, we could group by "kind" first. So we'd have

- Dtypes
  - CategoricalDtype
  - PeriodDtype
  - ...
- Scalars
  - Period
  - ...
- Arrays
   - Categorical
   - PeriodArray
   - ...

Do people have a preference between "by topic" and "by kind" (cc. @jorisvandenbossche @jreback @jbrockmendel @datapythonista)?

jorisvandenbossche commented:
> I'd like to restructure the documentation a bit.

I suppose you are only talking about the api.rst page?

TomAugspurger commented:
Yes, sorry.

TomAugspurger committed:
wip

jorisvandenbossche commented:
Not a strong opinion, but for adding new docs I would go by kind. 
But that is in the assumption that the above is only for listing the arrays / dtypes, and that there are still separate sections about each topic anyway? (eg with all Interval-specific or datetime-like methods/attributes)

TomAugspurger commented:
The nice thing about "by topic" is that we can give a high-level summary of what, e.g. "Period" is for.  If we have things grouped "by kind" (dtypes, scalars, arrays), then we'd need to repeat that description, or just have it once.

jorisvandenbossche commented:
> The nice thing about "by topic" is that we can give a high-level summary of what, e.g. "Period" is for

But in those places you would also put all the custom attributes and methods? 

I actually think we can be duplicated somewhat. Even if there are topical sections, I think it is still nice to have a short table of all dtypes and one for all array types.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
inference

TomAugspurger committed:
ia updates

TomAugspurger commented:
Did a bit of inference in fe06de4. There are a few TODOs.

I think we don't handle intervals yet. https://github.com/pandas-dev/pandas/issues/23553

TomAugspurger committed:
test fixup

TomAugspurger committed:
isort

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,84 @@
+def array(data, dtype=None, copy=False):
+    """
+    Create an array.
+
```
versionsdded tag

jreback commented on setup.cfg:
```diff
@@ -90,7 +90,7 @@ known_post_core=pandas.tseries,pandas.io,pandas.plotting
 sections=FUTURE,STDLIB,THIRDPARTY,PRE_CORE,DTYPES,FIRSTPARTY,POST_CORE,LOCALFOLDER
 
 known_first_party=pandas
-known_third_party=Cython,numpy,python-dateutil,pytz,pyarrow
```
can u remove anything u isorted

TomAugspurger replied:
I didn't isort anything (other than the new files, which aren't allowed to fail).

For whatever reason, my env running `isort` thought that that `pytest` was a first party module, so it grouped it with pandas.

TomAugspurger committed:
fixups

jreback commented on doc/source/api.rst:
```diff
@@ -790,6 +803,56 @@ following usable methods and properties:
+:class:`IntervalArray` is an array for storing data representing intervals.
+The scalar type is a :class:`Interval`. These may be stored in a :class:`Series`
+or as a :class:`IntervalIndex`. :class:`IntervalArray` can be closed on the
+left, right, or both, or neither sides.
```
use single back ticks on the closed; can you add a pointer to the interval docs themselves.

jreback commented on doc/source/api.rst:
```diff
@@ -790,6 +803,56 @@ following usable methods and properties:
+.. autosummary::
+   :toctree: generated/
+
+   PeriodArray
```
pointer to Period docs

jreback commented on doc/source/api.rst:
```diff
@@ -790,6 +803,56 @@ following usable methods and properties:
+
+.. _api.arrays.sparse:
+
+.. autosummary::
```
pointer to sparse docs

jreback commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -101,6 +101,18 @@ Reduction and groupby operations such as 'sum' work.
+
+.. ipython:: python
+
+   pd.array([1, 2, np.nan], dtype='Int64')
```
would show this first example w/o the dtype as well

jreback commented on pandas/arrays/__init__.py:
```diff
@@ -0,0 +1,17 @@
+)
+
+
+__all__ = [
```
we should consider renaming ``Categorical`` -> ``CategoricalArray`` (though deprecating maybe be very difficult)

TomAugspurger replied:
I think not worth it at this point.

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,86 @@
+
+from pandas.core.dtypes.common import is_extension_array_dtype
+from pandas.core.dtypes.dtypes import registry
+from pandas.core.dtypes.generic import ABCIndexClass, ABCSeries
```
do we have sufficient testing on EA's that raise on non 1-d input?

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,86 @@
+    Examples
+    --------
+    If a dtype is not specified, `data` is passed through to
+    :meth:`numpy.array`, and an ndarray is returned.
```
single backticks on ndarray

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,86 @@
+    If a dtype is not specified, `data` is passed through to
+    :meth:`numpy.array`, and an ndarray is returned.
+
+    >>> pd.array([1, 2])
```
can you show with a np.nan and no inference (and maybe with Int32) for the same

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,86 @@
+        return cls._from_sequence(data, dtype=dtype, copy=copy)
+
+    if dtype is None:
+        inferred_dtype = lib.infer_dtype(data)
```
use is_period_arraylike here (though IIRC we are trying to get rid of this), its the correct way for now (I think), see my next comment though

jreback replied:
I think

``lib.is_period_array(np.array(data, copy=False))`` might be the right way to do this (and will be *much* faster)

TomAugspurger replied:
Hmm, we only know that we have a `Sequence` at the point, not an ndarray like is expected by `is_period_array`.

I also suspect that we'll have more of these `elif`s for more dtypes, in which case maybe inferring the dtype once is faster than checking each? But perhaps not.

jreback commented on pandas/tests/arrays/test_array.py:
```diff
@@ -0,0 +1,109 @@
+import decimal
+
+import numpy as np
+import pytest
```
shouldn't this be in tests/extensions? and part of the extension base infrastructure?

TomAugspurger replied:
This is testing `pd.array` itself, not something that's part of the interface.

I did add a couple tests in `base/constructors.py` to check that `pd.array` is idempotent for EAs.

jreback commented on pandas/tests/arrays/test_array.py:
```diff
@@ -0,0 +1,109 @@
+     pd.IntervalArray.from_tuples([(1, 2), (3, 4)])),
+    ([0, 1], 'Sparse[int64]', pd.SparseArray([0, 1], dtype='int64')),
+    ([1, None], 'Int16', integer_array([1, None], dtype='Int16')),
+
```
can you also send in pd.Series / pd.Index of these types (or is this tested below)?

jreback replied:
do you test pass thru to ndarray? (e.g. maybe use any_numpy_dtype fixture with the dtype specified)

TomAugspurger replied:
It is tested below. Added a couple here as well though.

Not sure how `any_numpy_dtype` would be used here. We would need data to go with it.

jreback replied:
@TomAugspurger right you would need to construct a dummy array and test that its passing in thru

TomAugspurger replied:
Still not sure what this would look like, or what the test is for. I'm not especially worried about specific numpy types not getting through, and we do test that path.

jreback commented on pandas/tests/arrays/test_period.py:
```diff
@@ -5,15 +5,23 @@
+# Dtype
 
+
+def test_registered():
```
should this be an extension test? (or maybe in test_dtypes below)

jreback commented:
looks good @TomAugspurger mostly some docs & clarification comments, generally looks good though.

TomAugspurger commented:
From https://github.com/pandas-dev/pandas/pull/23581#discussion_r232490269

> https://github.com/pandas-dev/pandas/pull/23581#discussion_r232490269

No, we don't because I hadn't really considered what to do about that. Do people have thoughts on how to handle 2-D (or n-d) input inside `pd.array`? In general, it shouldn't be up to `pd.array` what is and isn't valid input. That's up to the individual array constructors. But is 2-D special since NumPy handles it but EAs don't?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger commented:
Another question: when given a series, index, or EA for `data`, we ignore `data.dtype` even if the user-specified dtype is `None`. So we get this (maybe strange?) result:

```python
In [2]: arr = pd.Categorical(['a', 'a', 'b'])

In [3]: pd.array(arr)
Out[3]: array(['a', 'a', 'b'], dtype=object)
```

Note, however, that we *cannot* just take a `dtype`  from an ndarray.  In the following, we ignore `data.dtype` (object) so that we can infer period.

```python
In [6]: arr = np.array([pd.Period('2000', 'D'), pd.Period('2000', 'D')])

In [7]: pd.array(arr)
Out[7]:
<PeriodArray>
['2000-01-01', '2000-01-01']
Length: 2, dtype: period[D]
```

So my proposal: when the user-specified `dtype` is None, pandas will

1. use the dtype from a Series / Index / EA, as if it was user-specfied. else
2. infer the the dtype from the data (including for ndarray)

TomAugspurger committed:
wip

jreback commented:
sounds reasonable for https://github.com/pandas-dev/pandas/pull/23581#issuecomment-437869870

TomAugspurger committed:
dtype from ea

TomAugspurger committed:
series, index tests

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
added ndarray case

jorisvandenbossche commented:
> So my proposal: when the user-specified dtype is None, pandas will
>
>    use the dtype from a Series / Index / EA, as if it was user-specfied. else
>    infer the the dtype from the data (including for ndarray)


I think the `pd.array(arr)` where `arr` is a Categorical returning not a categorical is something we should fix. 
But as I understand, that should be changed with above logic?

TomAugspurger commented:
Yes,
838ce5e makes pd.array idempotent for ExtensionArrays. We extract dtype info from Series, Index or EAs when `dtype=None`

TomAugspurger commented:
Revisiting this today. I think the biggest outstanding question is what to do about 2-D (or larger) input arrays.

I think attempting to detect 2-D arrays *before* they're created (from the raw `data`) is impossible. Especially if pandas or some 3rd party array wants to support nested data.

If so, our only option is to raise when we detect a 2D array output, which at this point just means an ndarray. Do we want that?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

jreback commented:
> If so, our only option is to raise when we detect a 2D array output, which at this point just means an ndarray. Do we want that?

I think we have checks in the pandas EA's for non-1D input. So if someone passes something non-1D and an EA dtype it won't return an ndarray (it will just raise), right?

can you show a case where there is a problem?

TomAugspurger commented:
> I think we have checks in the pandas EA's for non-1D input. So if someone passes something non-1D and an EA dtype it won't return an ndarray (it will just raise), right?

Right.

There's no real problem with pd.array returning 2-D arrays

```python
In [2]: pd.array([[1, 2], [3, 4]])
Out[2]:
array([[1, 2],
       [3, 4]])
```

it's just a matter of setting expectations. I'm fine with letting that pass through.

jreback commented:
@TomAugspurger yeah i agree, its either an EA, which may (or actually may not restrict) 2-Dness (e.g. imagine a Image EA array), or it passes to np.array

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
added test for a 2d array

TomAugspurger committed:
TST: test for Series[EA]

TomAugspurger commented:
Pushed some tests with support for 2D.

Test failure is the rank segfault.

This should be good to go.

jreback commented:
will look soon

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,113 @@
+
+        If not specified, there are two possibilities:
+
+        1. When `data` is a :class:`Series`, :class:`Index`, or
```
did you add tests for cases where you specify an EA (either thru Series, or directoy) and a dtype is specified? This could be a *different* dtype, what do we do then? .astype? (I think we should), e.g. this is analagous to

```
In [1]: arr = np.array([1,2])

In [2]: np.array(arr, dtype=float)
Out[2]: array([1., 2.])
```

we should mention this in the doc-string and test at least somewhat for this.

TomAugspurger replied:
I added a test.

I'm not sure it's worth mentioning in the docstring. It feels like a corner case, and is covered by the first line of the dtype description: `The dtype to use for the array. `

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,113 @@
+        inferred_dtype = lib.infer_dtype(data)
+        if inferred_dtype == 'period':
+            try:
+                return period_array(data)
```
should pass copy= here

jreback commented on pandas/tests/arrays/test_array.py:
```diff
@@ -0,0 +1,118 @@
+
+import pandas as pd
+from pandas.api.extensions import register_extension_dtype
+from pandas.core.arrays import integer_array, period_array
```
is decimal and json mentioned as first class dtypes? e.g. in the api? should we?

TomAugspurger replied:
Those are just test classes. They aren't part of the API.

TomAugspurger replied:
I think an arrow-backed Decimal array would make sense (this one is backed by an object-dtype ndarray).

jreback replied:
sure arrow backed things make sense, but until kernels are fully available an object based one is just nice to have. (and even then could just be a backend swap)

jorisvandenbossche commented:
Will try to do a review on Monday evening

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
Added test for period -> category

TomAugspurger committed:
copy

TomAugspurger committed:
prefix for arrays

TomAugspurger committed:
Added arrays

jorisvandenbossche commented:
> I think an arrow-backed Decimal array would make sense (this one is backed by an object-dtype ndarray).

I would leave that for a dedicated project for now (like fletcher), we already have a lot on our plate.

But that said: assume we add a DecimalArray (backed by arrow or not) in one of the next releases. How do we then see the "backwards compatibility guarantees" of this function? Because we would want to start returning a DecimalArray instead of object np.ndarray when decimals are passed to `pd.array`. But, this is of course not strictly backwards compatible. 
Assuming that situation, what should we do?

TomAugspurger commented:
> Assuming that situation, what should we do?

I think it comes down to the dtypes registry. All we can say is that for a given registry, `pd.array(..., dtype=dtype)` will be backwards compatible, right?

We could perhaps give users an API for manipulating the dtypes registry.

jorisvandenbossche commented:
> All we can say is that for a given registry, pd.array(..., dtype=dtype) will be backwards compatible, right?

That makes sense to me, but then we should make that clearer in the docstring? Because it basically means we guarantee no exact backwards compatibility if `dtype` is not specified.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
update docstring

TomAugspurger commented:
Added a note to that @jorisvandenbossche 

```
    Omitting the `dtype` argument means pandas will attempt to infer the
    best array type from the values in the data. As new array types are
    added by pandas and 3rd party libraries, the best array type may
    change. We recommend specifying `dtype` to ensure that the correct
    array type is constructed.
```

TomAugspurger committed:
docstring order

TomAugspurger committed:
Revert "docstring order"

This reverts commit c818a8f355695939a5bd87970c2738c29ce663f9.

jorisvandenbossche commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -106,6 +106,26 @@ Reduction and groupby operations such as 'sum' work.
+.. _whatsnew_0240.enhancements.array:
+
+A new top-level method :func:`array` has been added for creating arrays (:issue:`22860`).
+This can be used to create NumPy arrays, or any :ref:`extension array <extending.extension-types>`, including
```
Maybe a nitpick, but I would switch the order, to highlight its use to create extension arrays as its main purpose (I don't think people should start using it for just creating numpy arrays)

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,129 @@
+        2. Otherwise, pandas will attempt to infer the `dtype`
+           from the data.
+
+        Note that when `data` is a NumPy array, ``data.dtype`` is
```
this is not true, it is passed onto np.array

TomAugspurger replied:
This is trying to convey the fact that

```python
np.array([pd.Period('2000', 'D')], dtype=object)
```

will return a PeriodArray, rather than an object-dtype ndarray. I'll try to clarify that an ndarray's `.dtype` is ignored *for inference*.

jreback replied:
right, ok

jorisvandenbossche commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,129 @@
+        When `data` is an Index or Series, the underlying array
+        will be extracted from `data`.
+
+    dtype : Union[str, np.dtype, ExtensionDtype], optional
```
Do we want to use such type annotation terminology in user facing docstrings? I am fine with having them already as type comments, but using them in docstrings in general requires some discussion I think (to say, it's not in the current docstring guidelines). 
I think not everybody will now what we mean here with "Union"

jreback replied:
I think this is pretty common nowadays and pretty useful information.

jorisvandenbossche replied:
For people who don't see themselves as programmers (=most of our users) and who only interact with numpy, pandas, matplotlib, ... docstrings, I don't think this is "common"

(and it is also not really giving extra information, as "str, np.dtype or ExtensionDtype" also says "one of those three".

TomAugspurger replied:
All change to prose for now, but I think we'll want to do some educating of users here.

jorisvandenbossche commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,129 @@
+
+        1. When `data` is a :class:`Series`, :class:`Index`, or
+           :class:`ExtensionArray`, the `dtype` will be taken
+           from the data.
```
also when `data` is a numpy array I assume?

jorisvandenbossche replied:
Ah, I see the note a couple of lines below. However, I don't think that is completely true:

```
In [1]: pd.array(np.array([1, 2], dtype='int32'))                                                                                                                                                                   
Out[1]: array([1, 2], dtype=int32)

In [2]: pd.array([1, 2])                                                                                                                                                                                            
Out[2]: array([1, 2])
```

Maybe a logical example, but so strictly spoken, `data.dtype` is not ignored here in that first case.

TomAugspurger replied:
I've clarified that `data.dtype` is ignored only when it comes to inference.

jreback reviewed (changes requested):
doc comment

jorisvandenbossche commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,129 @@
+        Note that when `data` is a NumPy array, ``data.dtype`` is
+        ignored.
+
+    copy : bool, default False
```
Numpy has a default of True. Do we have a rationale for using False by default?

TomAugspurger replied:
Was just matching series here. I don't have a strong preference. Let's just match NumPy then.

jorisvandenbossche commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,129 @@
+    Omitting the `dtype` argument means pandas will attempt to infer the
+    best array type from the values in the data. As new array types are
+    added by pandas and 3rd party libraries, the best array type may
+    change. We recommend specifying `dtype` to ensure that the correct
```
```suggestion
    change. We recommend specifying `dtype` to ensure that a stable and the correct
```

My suggestion is maybe not fully a proper sentence construction, but wanted to have some notion in it about if you want a stable result across versions. 
(but thanks for adding the note!)

jreback commented:
lgtm. ping on green.

jorisvandenbossche commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,129 @@
+           :class:`ExtensionArray`, the `dtype` will be taken
+           from the data.
+        2. Otherwise, pandas will attempt to infer the `dtype`
+           from the data.
```
I find this statement a bit misleading, as we actually don't infer from scalars (at least not in the sense of eg how Series does it), we only use numpy's inference? 

If we say that we infer, I would expect those to do the same:

```
In [3]: pd.array([pd.Timestamp("2012-01-01")])                                                                                                                                                                      
Out[3]: array([Timestamp('2012-01-01 00:00:00')], dtype=object)

In [5]: pd.array([pd.Timestamp("2012-01-01")], dtype='datetime64[ns]')                                                                                                                                              
Out[5]: array(['2012-01-01T00:00:00.000000000'], dtype='datetime64[ns]')
```

jorisvandenbossche replied:
Ah, looking now further down in the implementation :-) I see you do handle this case for Period, so here it indeed infers:

```
In [6]: pd.array([pd.Period('2012-01-01', freq='D')])                                                                                                                                                               
Out[6]: 
<PeriodArray>
['2012-01-01']
Length: 1, dtype: period[D]
```

TomAugspurger replied:
I haven't stated it explicitly, but it would be nice if 3rd parties could eventually hook into this as well. Right now I think it's just Period (and maybe interval?) that get inferred. Maybe timestamps with timezones once DatetimeArray is done.

jorisvandenbossche replied:
And not Timestamps without timezones?

TomAugspurger replied:
It seems that `lib.infer_dtype` doesn't distinguish the two

```python
In [5]: lib.infer_dtype([pd.Timestamp('2017', tz='utc')])
Out[5]: 'datetime'

In [6]: lib.infer_dtype([pd.Timestamp('2017', tz='US/Central')])
Out[6]: 'datetime'
```

I've added interval, and tests.

jorisvandenbossche replied:
> Perhaps we should be explicit in the docs for that? 

Yes, I would do that. 
Would we, in the longer term, the inference happening here be the same as the inference happening in `Series` ? Or would that actually add to much corner cases from there we don't want to carry over?

> Though that kinda closes the door to inference for 3rd party arrays.

How would you envision third party arrays participate in inference? That seems a bit difficult in any case (trying out all registered ones, ..?), and IMO more error prone for users (if you forget to import the 3rd party library, you silently get different results)

jorisvandenbossche commented:
> lgtm. ping on green.

To be explicit about it this time, don't merge it yet on green, I am still reviewing it.

jreback commented:
@jorisvandenbossche ok sure. in the future, easiest if you simply put a review block then.

jorisvandenbossche reviewed (commented):
Only got until the docstring, have a meeting now, but will continue after that.

One thing to maybe already consider for the upcoming DatetimeArray: what do we do with `dtype='datetime64[ns]'` ? Return a DatetimeArray or np.ndarray? 

(I was actually thinking we could maybe only allow strings for extension dtypes, and not as alias for numpy dtypes, to make it more backwards compatible in the future)

TomAugspurger committed:
Updates

* interval inference
* copy by default
* doc clarifications

jreback commented:
@jorisvandenbossche if you have any more comments.

jreback commented:
@jorisvandenbossche

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
Add docs for the types we infer

TomAugspurger commented:
Added a note to the docs about the extension types that we infer.

jreback commented:
lgtm. let's merge this.

jorisvandenbossche commented:
@TomAugspurger BTW, thanks for the earlier updates of a few days ago. The added note is exactly what I was thinking about.

Any comment on this:

> (I was actually thinking we could maybe only allow strings for extension dtypes, and not as alias for numpy dtypes, to make it more backwards compatible in the future)

?

(we can certainly discuss that further after merging, it doesn't need to hold up the PR if it is blocking other things)

jorisvandenbossche reviewed (changes requested):
Not yet sure if it is related to the changes in this PR, but the doc build completely failed on the api section.

TomAugspurger commented:
I'm chewing that over right now... The future concerns are

1. We or a 3rd party adds an extension dtype whose alias is the same as an ndarray. (datetime64[ns])
2. We want to change the default "fallback" away from numpy in the future, so that `pd.array([1, 2], dtype='int64')` returns, say, an Apache Arrow array.

I *think* that 2 is disruptive enough, that we would have some kind of separate parameter for that, right?

I'm really not sure on this... but I think that raising on aliases passed to numpy is a good idea for future proofing... It's less convenient, but we can have a nice error message.

So the proposal is that

```python
pd.array([1, 2], dtype='int64')
```

would raise with a message like

```python
ValueError("Ambiguous 'dtype': Pandas doesn't pass string aliases to NumPy. Specify `dtype=np.dtype('int64')` instead.)
```
```

jbrockmendel commented:
>  the inference happening here be the same as the inference happening in `Series` ?

Along with #23623, would it make sense for pandas.array to just be `return Series(data, dtype=dtype).array`?  Maybe not _that_ simple, but there should be a pretty close correspondence right?

TomAugspurger commented:
Perhaps... though I'd rather that `Series.__init__` called `pd.array` internally, to avoid doing unnecessary work in `pd.array`.

TomAugspurger committed:
API: disallow string alias for NumPy

TomAugspurger commented:
https://github.com/pandas-dev/pandas/pull/23581/commits/94064009258259cd9821bab2c07db95002108408 disallows passing string aliases to `numpy.array(..., dtype=dtype)` from within `pd.array()`. Hopefully things make sense.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
Wrap long error message

jreback commented:
> Are we OK with pd.array not allowing NumPy string aliases? That introduces an inconsistency with Series, which is unfortunate.

why would we do that? I think it should just pass things thru

TomAugspurger commented:
@jorisvandenbossche I think I agree with Jeff about the string aliases. The inconsistency with Series is unfortunate. I will keep a section in the docs noting that possibility though.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
updates

* Removed raising on strings
* Docstring passes

jreback commented:
> The inconsistency with Series is unfortunate. 

what do you mean here @TomAugspurger ?

TomAugspurger commented:
Sorry, the inconsistency of `pd.array([1, 2], dtype='int')` raising, while `pd.Series([1, 2], dtype='int')` didn't. With the latest commit they're consistent (pass through to `np.dtype`).

jreback commented:
> Sorry, the inconsistency of pd.array([1, 2], dtype='int') raising, while pd.Series([1, 2], dtype='int') didn't. With the latest commit they're consistent (pass through to np.dtype).

ahh ok, yes I fully agree with this; its the least suprising. And I don't even thing this is a backwards compat issue that we should even try to address now.

TomAugspurger committed:
removed old test

jreback commented on pandas/tests/arrays/test_array.py:
```diff
@@ -0,0 +1,158 @@
+    (np.array([1, 2]), None, np.array([1, 2])),
+    # String alias passes through to NumPy
+    ([1, 2], 'float32', np.array([1, 2], dtype='float32')),
+    # Period alias
```
I would just put a blank line between cases, I think that helps a lot (e.g. before the comment)

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
formatting

TomAugspurger commented:
@jorisvandenbossche do you want us to hold off on merging this? Any short feedback on not implementing the "raise on string alias passed to NumPy?"

jreback commented:
let's merge this. need to whittle down the queue. @jorisvandenbossche certainly can have followups.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger commented:
Fixed the merge conflict. @jorisvandenbossche are you OK with passing string aliases through to `np.array` rather than raising?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

jreback commented:
this has been hanging way too long. @jorisvandenbossche final look today, otherwise merging. if you have comments they can be addressed in a followup.

jorisvandenbossche commented:
Yeah, as I mentioned above, the discussion on accepting string aliases or not doesn't necessarily need to hold up merging the *PR*. 

But, I am still not sure we should support it. It's true that there is an inconsistency with Series on that regard, but you could see `pd.array` as a little bit lower level as Series, where you first need to process the dtype. To illustrate it with a hypothetical series constructor (ignoring a lot of things of course):

```
def __init__(self, data, index, dtype=None):
    if dtype is not None:
        dtype = pandas_dtype(dtype)
    data = pd.array(data, dtype=dtype)
    ....
```

Or, we could also simply not return any numpy array at all, and only restrict the function for now to be a public constructor for our own ExtensionArrays. 
For me, the question is: what do people want to use this for / in what usage patterns? Just to manually create a PeriodArray or IntegerArray? In that case, returning only EAs might be enough (in the idea of: better keep the scope restricted, we can always expand it if there is demand for it)
(my main concern is: how can we prevent newcomers to start using `pd.array` when they actually should use `np.array(...)`)

jorisvandenbossche commented:
> (my main concern is: how can we prevent newcomers to start using pd.array when they actually should use np.array(...))

@TomAugspurger and thanks for the additional note you added on recommending to not use string aliases, that certainly addresses this concern somewhat (now we only need users to read the docs ... :-))

jorisvandenbossche reviewed (changes requested):
Regardless of the other discussion on what types to allow / to return, I think that, when passing things through to numpy, we should at least restrict the result to be 1D:

```
In [1]: pd.array([[1, 2], [3, 4]])                                                                                                                                                                                  
Out[1]: 
array([[1, 2],
       [3, 4]])
```

So I would raise in this case?

TomAugspurger commented:
> For me, the question is: what do people want to use this for / in what usage patterns?

Yeah thats' hard to say at this point.

> Just to manually create a PeriodArray or IntegerArray? 

To be clear, we'll add DatetimeArray to this list.

> better keep the scope restricted, we can always expand it if there is demand for it)

Indeed. Or mark it as experimental :)

> (my main concern is: how can we prevent newcomers to start using pd.array when they actually should use np.array(...))

There's a symmetry with @shoyer's proposal for `Series.array` to always be an ExtensionArray here. If we wrote a small NumpyExtensionArray wrapper, that would be the natural thing to return here. And if we did that 2d input would be out...

jreback commented:
yeah I am ok with no allowing non 1-D input (would scalars be allowed as well?).

I guess we want this to generally expose ExtensionArray construction w/o having to use PeriodArray for example directly.

maybe call this ``.to_array()`` instead? (along the lines of our ``.to_timedelta, to_datetime, to_numeric`` ?

TomAugspurger commented:
Oh I haven't thought at all about scalars. I assume we want to raise there? Will add tests.

Don't have a strong preference between `to_array` and `array`. I think since both NumPy and pyarrow call this method `array` then I'm leaning towards that.

Assuming we decide to make `Series.array` always an ExtensionArray, do people agree that `pd.array` always returns an extension array?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
Raise on scalars

TomAugspurger commented:
Added a test to raise on scalars.

Thinking on it more, I think we should implement @shoyer's proposal of Series.array always being an ExtensionArray. That's easy to do and gives us a lot of freedom going forward. In that spirit, this should return an ExtensionArray only, and not fall back to numpy.

Will have an update later today.

jorisvandenbossche commented:
> Assuming we decide to make Series.array always an ExtensionArray, do people agree that pd.array always returns an extension array?

Yes, I think it makes sense to have those two aligned. 
I personally like the idea of always an ExtensionArray, but also don't know how much effort that would be? We shouldn't add too much code checks throughout our code base just for this (because we would handle those ExtensionArrays differently than others, no? Eg when passing a "NumpyBasedExtensionArray" to `Series(..)` we want to create a "normal" Series and not one backed by an actual ExtensionArray using the interface?

jorisvandenbossche commented on pandas/core/arrays/array_.py:
```diff
@@ -147,6 +147,12 @@ def array(data,         # type: Sequence[object]
+        msg = (
+            "Cannot pass scalar '{}' to 'pandas.array'."
+        )
+        raise ValueError(msg.format(data))
```
You could also combine this check with checking in the end that the returned value has a dimension of 1 (so not 0 for scalars or >= 2)

TomAugspurger replied:
I think we should have it early, to avoid passing scalars to `ExtensionArray._from_sequence`. Subclasses should probably validate that on their own, but I think we'd like a consistent error message.

TomAugspurger commented:
> I personally like the idea of always an ExtensionArray, but also don't know how much effort that would be? We shouldn't add too much code checks throughout our code base just for this (because we would handle those ExtensionArrays differently than others, no?

@jorisvandenbossche a few comments on that starting at 
https://github.com/pandas-dev/pandas/issues/23995#issuecomment-445956871

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -158,6 +158,28 @@ Reduction and groupby operations such as 'sum' work.
+A new top-level method :func:`array` has been added for creating arrays (:issue:`22860`).
+This can be used to create any :ref:`extension array <extending.extension-types>`, including
+extension arrays registered by :ref:`3rd party libraries <ecosystem.extensions>`, or to
+create NumPy arrays.
```
specify this as 1D here?

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,190 @@
+    Pandas will infer an ExtensionArray for some types of data:
+
+    >>> pd.array([pd.Period('2000', freq="D"), pd.Period("2000", freq="D")])
+    <PeriodArray>
```
can you add an example which raises for scalars, 2d?

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
docs on raising

TomAugspurger commented:
Added a doc note about just doing 1-d arrays.

Holding off on NumPy / 2D changes for now, in case we decide we like `Series.array` always being an EA (will have a PR in an hour or so).

jreback commented:
> But, specific to the pd.array function If we go with #24227, would we return a NumPyBackedExtensionArray, or raise, for cases like

pd.array([1, 2, 3])

I guess you have to raise on this until we have numpy backed EA (which I don't think we should try to put into 0.24.0). 

Though I am ok with just returning a numpy array for now.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
Updates for PandasArray

TomAugspurger commented:
932e119c3383400e78d733f713acd76ae837b765 has the changes for `PandasArray`. API-wise, this means that `pd.array` always returns an ExtensionArray.

This implies that `pd.array` raises for non-1d input (scalars or 2+ dimensions).

TomAugspurger committed:
update docstring

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -161,6 +161,39 @@ Reduction and groupby operations such as 'sum' work.
+
+A new top-level method :func:`array` has been added for creating 1-dimensional arrays (:issue:`22860`).
+This can be used to create any :ref:`extension array <extending.extension-types>`, including
+extension arrays registered by :ref:`3rd party libraries <ecosystem.extensions>`.
```
don't you now have a ref in basics where this should point?

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,231 @@
+        )
+        raise ValueError(msg.format(data))
+
+    if isinstance(data, (ABCSeries, ABCIndexClass)):
```
extract_array?

jreback commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,231 @@
+        # TODO(BooleanArray): handle this type
+
+    result = np.array(data, dtype=dtype, copy=copy)
+    result = PandasArray(result)
```
this takes copy= now right? (maybe we should *also* take dtype?); I am talking into PandasArray

TomAugspurger replied:
right, `PandasArray.__init__` takes `copy` now.

I think taking `dtype` isn't worthwhile.  Right now `PandasArray` doesn't call `np.asarray` or `np.array` on the input. If we accepted `dtype` we would need to call one of those. What would be your motivation for accepting in?

jreback replied:
its fine, was just trying to remove the need to do the coercion here (and instead do inside PandasArray), but this looks ok, though maybe ``PandasArray._from_sequence``?

TomAugspurger replied:
I see. `PandasArray._from_sequence` makes sense.

jreback reviewed (changes requested):
small comments

TomAugspurger committed:
Updates

* doc ref
* use extract_array
* use PandasArray._from_sequence

jreback commented:
@TomAugspurger I am ok with merging this on green. Can follow up from the prior @jorisvandenbossche comment, which was about how we are handling string dtypes I think? can you create an issue for that discussion

TomAugspurger commented:
Thanks for the review.

I think that https://github.com/pandas-dev/pandas/pull/23581/files#diff-69ac57923b848af43df327c311b79db4R90 handles @jorisvandenbossche's comments regarding string aliases for dtypes. In a world where we have a `StringArray` backed by apache arrow

```python
pd.array(['a', 'b'], dtype=None/str)
```

would return a StringArray. But

```python
pd.array(['a', 'b'], dtype=np.dtype(str))
```

would continue to return a `PandasArray` backed by an ndarray with dtype `np.dtype("<U")`.

We should maybe emphasize more that if the underlying memory layout *really* matters to you, then you shouldn't be using `pd.array`. But I think this is a relatively rare case, and don't want to bog down users with low-level details...

jreback commented:
https://github.com/pandas-dev/pandas/pull/23581#issuecomment-450418326 sounds ok to me.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger committed:
fixed test expected

TomAugspurger committed:
doc lint

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pd.array

TomAugspurger commented:
All green now.

jreback commented:
awesome as always @TomAugspurger

jbrockmendel commented on pandas/core/arrays/array_.py:
```diff
@@ -0,0 +1,227 @@
+
+    # this returns None for not-found dtypes.
+    if isinstance(dtype, compat.string_types):
+        dtype = registry.find(dtype) or dtype
```
@TomAugspurger do you remember if there was any particular reason for using this pattern instead of `dtype = pandas_dtype(dtype)`?

TomAugspurger replied:
I wonder if this predates `pandas_dtype` handling extension dtypes.
